infra/bisector.py

```diff
@@ -113,40 +111,52 @@ def bisect(commit_old, commit_new, testcase, fuzz_target, build_data):
+    ValueError: when a repo url can't be determine from the project
   """
-  local_store_path = tempfile.mkdtemp()
-  repo_url = build_specified_commit.infer_main_repo(build_data.project_name,
-                                                    local_store_path,
-                                                    commit_old)
-  bisect_repo_manager = repo_manager.RepoManager(repo_url, local_store_path)
-  commit_list = bisect_repo_manager.get_commit_list(commit_old, commit_new)
-  build_specified_commit.build_fuzzer_from_commit(
-      build_data.project_name, commit_list[0], bisect_repo_manager.repo_dir,
-      build_data.engine, build_data.sanitizer, build_data.architecture,
-      bisect_repo_manager)
-  error_code = helper.reproduce_impl(build_data.project_name, fuzz_target,
-                                     False, [], [], testcase)
-  old_idx = len(commit_list) - 1
-  new_idx = 0
-  if len(commit_list) == 1:
-    if not error_code:
-      return None
-    return commit_list[0]
-
-  while old_idx - new_idx != 1:
-    curr_idx = (old_idx + new_idx) // 2
+  with tempfile.TemporaryDirectory() as tmp_dir:
+    repo_url, repo_name = build_specified_commit.detect_main_repo_from_docker(
+        build_data.project_name, commit_old)
+    if not repo_url or not repo_name:
+      raise ValueError('Main git repo can not be determined.')
+    bisect_repo_manager = repo_manager.RepoManager(
+        repo_url, tmp_dir, repo_name=repo_name)
+    commit_list = bisect_repo_manager.get_commit_list(commit_old, commit_new)
+    old_idx = len(commit_list) - 1
+    new_idx = 0
+    build_specified_commit.build_fuzzer_from_commit(
+        build_data.project_name, commit_list[new_idx],
+        bisect_repo_manager.repo_dir, build_data.engine, build_data.sanitizer,
+        build_data.architecture, bisect_repo_manager)
+    expected_error_code = helper.reproduce_impl(build_data.project_name,
+                                                fuzz_target, False, [], [],
+                                                testcase)
+
+    # Check if the error is persistent through the commit range
     build_specified_commit.build_fuzzer_from_commit(
-        build_data.project_name, commit_list[curr_idx],
+        build_data.project_name, commit_list[old_idx],
         bisect_repo_manager.repo_dir, build_data.engine, build_data.sanitizer,
         build_data.architecture, bisect_repo_manager)
-    error_exists = (
-        helper.reproduce_impl(build_data.project_name, fuzz_target, False, [],
-                              [], testcase) == error_code)
-    if error_exists == error_code:
-      new_idx = curr_idx
-    else:
-      old_idx = curr_idx
-  return commit_list[new_idx]
+    oldest_error_code = helper.reproduce_impl(build_data.project_name,
+                                              fuzz_target, False, [], [],
+                                              testcase)
+    if expected_error_code == oldest_error_code:
+      return commit_list[old_idx]
```

Comment:
I think we should error out here with a message that "Both the first and the last commits in the given range have the same behavior, bisection is not possible."

Reply:
Done.